zerver/actions/user_settings.py

```diff
@@ -460,222 +462,244 @@ def do_change_user_setting(
         else:
             assert isinstance(db_setting_value, property_type)
             assert isinstance(event_value, property_type)
-    setattr(user_profile, setting_name, db_setting_value)
 
-    # TODO: Move these database actions into a transaction.atomic block.
-    user_profile.save(update_fields=[setting_name])
+    for user in user_profiles:
+        old_value = getattr(user, setting_name)
+        old_values[user.id] = old_value
+
+        setattr(user, setting_name, db_setting_value)
+
+        audit_logs.append(
+            RealmAuditLog(
+                realm=user.realm,
+                event_type=AuditLogEventType.USER_SETTING_CHANGED,
+                event_time=event_time,
+                acting_user=acting_user,
+                modified_user=user,
+                extra_data={
+                    RealmAuditLog.OLD_VALUE: old_value,
+                    RealmAuditLog.NEW_VALUE: db_setting_value,
+                    "property": setting_name,
+                },
+            )
+        )
 
-    RealmAuditLog.objects.create(
-        realm=user_profile.realm,
-        event_type=AuditLogEventType.USER_SETTING_CHANGED,
-        event_time=event_time,
-        acting_user=acting_user,
-        modified_user=user_profile,
-        extra_data={
-            RealmAuditLog.OLD_VALUE: old_value,
-            RealmAuditLog.NEW_VALUE: db_setting_value,
-            "property": setting_name,
-        },
-    )
+    UserProfile.objects.bulk_update(user_profiles, [setting_name])
+    RealmAuditLog.objects.bulk_create(audit_logs)
 
-    # Disabling digest emails should clear a user's email queue
-    if setting_name == "enable_digest_emails" and not db_setting_value:
-        clear_scheduled_emails([user_profile.id], ScheduledEmail.DIGEST)
+    for user in user_profiles:
+
+        def make_flush_function(user_profile: UserProfile) -> Callable[[], None]:
+            return lambda: flush_user_profile(sender=UserProfile, instance=user_profile)
 
-    if setting_name == "email_notifications_batching_period_seconds":
-        assert isinstance(old_value, int)
-        assert isinstance(db_setting_value, int)
-        update_scheduled_email_notifications_time(user_profile, old_value, db_setting_value)
+        transaction.on_commit(make_flush_function(user))
```

Comment:
This seems crazy; we should use a bulk version. We have `delete_user_profile_caches` for most of the work; I think you should be able to do a preparatory refactor PR to make `flush_user_profile` call a `bulk_flush_users` helper that, for example, only calls `delete_user_profile_caches` once and collects the things to delete for a single call to `cache_delete_many`.